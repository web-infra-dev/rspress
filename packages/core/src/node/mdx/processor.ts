import path from 'node:path';
import { createProcessor } from '@mdx-js/mdx';
import type { Header, UserConfig } from '@rspress/shared';
import { logger } from '@rspress/shared/logger';
import { extractTextAndId, loadFrontMatter } from '@rspress/shared/node-utils';

import type { PluginDriver } from '../PluginDriver';
import type { RouteService } from '../route/RouteService';
import {
  applyReplaceRules,
  escapeMarkdownHeadingIds,
  normalizePath,
} from '../utils';
import { createMDXOptions } from './options';
import type { TocItem } from './remarkPlugins/toc';
import type { PageMeta } from './types';

interface CompileOptions {
  source: string;
  filepath: string;
  checkDeadLinks: boolean;

  // assume that the below instances are singleton, it will not change.
  docDirectory: string;
  config: UserConfig | null;
  routeService: RouteService | null;
  pluginDriver: PluginDriver | null;
}

async function compile(options: CompileOptions): Promise<string> {
  const {
    source,
    filepath,
    docDirectory,
    checkDeadLinks,
    config,
    routeService,
    pluginDriver,
  } = options;

  const mdxOptions = await createMDXOptions({
    checkDeadLinks,
    config,
    docDirectory,
    filepath,
    pluginDriver,
    routeService,
  });
  // Separate frontmatter and content in MDX source
  const { frontmatter, emptyLinesSource } = loadFrontMatter(
    source,
    filepath,
    docDirectory,
    true,
  );

  // For loader error stack, so we use the source with frontmatter @see https://github.com/web-infra-dev/rspress/issues/2010
  const { replaceRules } = config ?? {};
  const replacedSource = applyReplaceRules(emptyLinesSource, replaceRules);

  // Support custom id like `#hello world {#custom-id}`.
  // TODO > issue: `#hello world {#custom-id}` will also be replaced.
  const preprocessedContent = escapeMarkdownHeadingIds(replacedSource);

  try {
    let pageMeta: PageMeta = {
      title: '',
      toc: [] as TocItem[],
      headingTitle: '',
    };

    const frontmatterTitle = extractTextAndId(frontmatter.title)[0];

    const compiler = createProcessor(mdxOptions);

    compiler.data('pageMeta' as any, { toc: [], title: '' });
    const vFile = await compiler.process({
      value: preprocessedContent,
      path: filepath,
    });

    const compileResult = String(vFile);
    const compilationMeta = compiler.data('pageMeta' as any) as {
      toc: Header[];
      title: string;
    };

    const headingTitle = extractTextAndId(compilationMeta.title)[0];
    pageMeta = {
      ...compilationMeta,
      title: frontmatterTitle || headingTitle,
      headingTitle,
      frontmatter,
    } as PageMeta;

    const result = `const frontmatter = ${JSON.stringify(frontmatter)};
${compileResult}
MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["${encodeURIComponent(
      normalizePath(path.relative(docDirectory, filepath)),
    )}"] = ${JSON.stringify(pageMeta)};
`;
    return result;
  } catch (e: unknown) {
    if (e instanceof Error) {
      throw e;
    }
    logger.error(`MDX compile error: ${e} in ${filepath}`);
    throw new Error(`MDX compile error: ${e} in ${filepath}`);
  }
}

// TODO: free the memory
const cache = new Map<string, Promise<string>>();

async function compileWithCrossCompilerCache(
  options: CompileOptions,
): Promise<string> {
  const task = cache.get(options.filepath);
  if (task) {
    return task;
  }

  const promise = compile(options);
  cache.set(options.filepath, promise);
  return promise;
}

export { compile, compileWithCrossCompilerCache };
