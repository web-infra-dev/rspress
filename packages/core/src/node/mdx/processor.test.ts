import { describe, it } from 'vitest';
import { compile } from './processor';

describe('mdx', () => {
  it('basic', async () => {
    const result = await compile({
      source: `
# Hello World
`,
      checkDeadLinks: false,
      docDirectory: '/usr/rspress-project/docs',
      filepath: '/usr/rspress-project/docs/index.mdx',
      config: null,
      pluginDriver: null,
      routeService: null,
    });
    expect(result).toMatchSnapshot();
  });

  it('should render inline code in title', async () => {
    const result = await compile({
      source: `
# Hello World \`inline code\`
`,
      checkDeadLinks: false,
      docDirectory: '/usr/rspress-project/docs',
      filepath: '/usr/rspress-project/docs/inline-code.mdx',
      config: null,
      pluginDriver: null,
      routeService: null,
    });
    expect(result).toMatchSnapshot();
  });

  it('should allow custom id', async () => {
    const result = await compile({
      source: `
# Hello World \`inline code\` \\{#custom-id}
`,
      checkDeadLinks: false,
      docDirectory: '/usr/rspress-project/docs',
      filepath: '/usr/rspress-project/docs/inline-code.mdx',
      config: null,
      pluginDriver: null,
      routeService: null,
    });
    expect(result).toMatchSnapshot();
  });
});
