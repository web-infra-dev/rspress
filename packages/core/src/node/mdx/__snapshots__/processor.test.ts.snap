// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mdx > basic 1`] = `
"const frontmatter = {};
import {jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
function _createMdxContent(props) {
  const _components = {
    a: "a",
    h1: "h1",
    ..._provideComponents(),
    ...props.components
  };
  return _jsxs(_components.h1, {
    id: "hello-world",
    children: [_jsx(_components.a, {
      className: "header-anchor",
      "aria-hidden": "true",
      href: "#hello-world",
      children: "#"
    }), "Hello World"]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = {
    ..._provideComponents(),
    ...props.components
  };
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}

MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["index.mdx"] = {"toc":[],"title":"Hello World","headingTitle":"Hello World","frontmatter":{}};
"
`;

exports[`mdx > should allow custom id 1`] = `
"const frontmatter = {};
import {jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
function _createMdxContent(props) {
  const _components = {
    a: "a",
    code: "code",
    h1: "h1",
    ..._provideComponents(),
    ...props.components
  };
  return _jsxs(_components.h1, {
    id: "custom-id",
    children: [_jsx(_components.a, {
      className: "header-anchor",
      "aria-hidden": "true",
      href: "#custom-id",
      children: "#"
    }), "Hello World ", _jsx(_components.code, {
      children: "inline code"
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = {
    ..._provideComponents(),
    ...props.components
  };
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}

MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["inline-code.mdx"] = {"toc":[],"title":"Hello World \`inline code\`","headingTitle":"Hello World \`inline code\`","frontmatter":{}};
"
`;

exports[`mdx > should render inline code in title 1`] = `
"const frontmatter = {};
import {jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
function _createMdxContent(props) {
  const _components = {
    a: "a",
    code: "code",
    h1: "h1",
    ..._provideComponents(),
    ...props.components
  };
  return _jsxs(_components.h1, {
    id: "hello-world-inline-code",
    children: [_jsx(_components.a, {
      className: "header-anchor",
      "aria-hidden": "true",
      href: "#hello-world-inline-code",
      children: "#"
    }), "Hello World ", _jsx(_components.code, {
      children: "inline code"
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = {
    ..._provideComponents(),
    ...props.components
  };
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}

MDXContent.__RSPRESS_PAGE_META = {};

MDXContent.__RSPRESS_PAGE_META["inline-code.mdx"] = {"toc":[],"title":"Hello World \`inline code\`","headingTitle":"Hello World \`inline code\`","frontmatter":{}};
"
`;
